research/object_detection/models/faster_rcnn_resnet_v1_fpn_keras_feature_extractor.py

```diff
@@ -0,0 +1,423 @@
+        outputs = [full_resnet_v1_model.get_layer(output_layer_name).output
+                   for output_layer_name in output_layers]
+        self.classification_backbone = tf.keras.Model(
+            inputs=full_resnet_v1_model.inputs,
+            outputs=outputs)
+
+        self._base_fpn_max_level = min(self._fpn_max_level, 5)
+        self._num_levels = self._base_fpn_max_level + 1 - self._fpn_min_level
+        self._fpn_features_generator = (
+            feature_map_generators.KerasFpnTopDownFeatureMaps(
+                num_levels=self._num_levels,
+                depth=self._additional_layer_depth,
+                is_training=self._is_training,
+                conv_hyperparams=self._conv_hyperparams,
+                freeze_batchnorm=self._freeze_batchnorm,
+                name='FeatureMaps'))
+
+        # Construct coarse feature layers
+        for i in range(self._base_fpn_max_level, self._fpn_max_level):
+          layers = []
+          layer_name = 'bottom_up_block{}'.format(i)
+          layers.append(
+              tf.keras.layers.Conv2D(
+                  self._additional_layer_depth,
+                  [3, 3],
+                  padding='SAME',
+                  strides=2,
+                  name=layer_name + '_conv',
+                  **self._conv_hyperparams.params()))
+          layers.append(
+              self._conv_hyperparams.build_batch_norm(
+                  training=(self._is_training and not self._freeze_batchnorm),
+                  name=layer_name + '_batchnorm'))
+          layers.append(
+              self._conv_hyperparams.build_activation_layer(
+                  name=layer_name))
+          self._coarse_feature_layers.append(layers)
+
+        feature_extractor_model = ResnetFPN(self.classification_backbone,
+                                            self._fpn_features_generator,
+                                            self._coarse_feature_layers,
+                                            self._pad_to_multiple,
+                                            self._fpn_min_level,
+                                            self._resnet_block_names,
+                                            self._base_fpn_max_level)
+        return feature_extractor_model
+
+  def get_box_classifier_feature_extractor_model(self, name=None):
+    """Returns a model that extracts second stage box classifier features.
+
+    Construct two fully connected layer to extract the box classifier features.
+
+    Args:
+      name: A scope name to construct all variables within.
+
+    Returns:
+      A Keras model that takes proposal_feature_maps:
+        A 4-D float tensor with shape
+        [batch_size * self.max_num_proposals, crop_height, crop_width, depth]
+        representing the feature map cropped to each proposal.
+
+      And returns proposal_classifier_features:
+        A 4-D float tensor with shape
+        [batch_size * self.max_num_proposals, 1024]
```

Comment:
correct output shape is `[batch_size * self.max_num_proposals, 1, 1, 1024]` right ?